### PR TITLE
Added support for multiple directories to deploy OOD extensions/customizations

### LINKED ISSUE #3916: OOD Customizations Deployment - Individual folder support
Currently, Open OnDemand allows customizations and extensions using three primary external folders:
 - `/etc/ood/config/apps/dashboard/initializers`
 - `/etc/ood/config/apps/dashboard/lib`
 - `/etc/ood/config/apps/dashboard/views`

This structure enables developers to create and integrate custom widgets, utilities, and other modifications to enhance Open OnDemand’s functionality and user experience. However, as more customizations are added, managing these additions becomes challenging.

To address this, we propose an enhancement to Open OnDemand that would support independent folders for each customization, each with its own `initializers`, `lib`, and `views` subdirectories. This modular approach would allow each customization to reside in its own self-contained structure, making it easier to add, remove, or share specific features. 

We believe that this deployment enhancement will better support sharing customizations among institutions in the Open OnDemand community.

### Proposed Structure
This structure would allow customizations to be packaged and shared as independent units:

```
/etc/ood/config/dashboard/add-ons
├── customization_1
│   ├── initializers
│   ├── lib
│   └── views
├── customization_2
│   ├── initializers
│   ├── lib
│   └── views
└── customization_3
    ├── initializers
    ├── lib
    └── views

```

Possible implementation pseudocode:
```ruby
# dashboard/config/application.rb
# plugin_directory = '/var/www/ood/apps/plugins'
...
    # Enable installed plugins
    if ::Configuration.plugin_directory.directory?
      ::Configuration.plugin_directory.children.select(&:directory?).each do |installed_plugin|
        config.paths["config/initializers"] << installed_plugin.join("initializers").to_s
        config.autoload_paths << installed_plugin.join("lib").to_s
        config.paths["app/views"].unshift installed_plugin.join("views").to_s
      end
    end
...
```

### PR DESCRIPTION
**Pull request by abujeda**
Experimental feature to add support for a folder based OOD extensions structure.
The idea being that administrators can deploy OOD extensions by copying a folder to the new plugins directory.

More infor about the idea in #3916

To test:
Create the plugins folder:` /var/www/ood/apps/plugins`
Download the attached zip file. It contains a folder with an OOD extension.
Add the folder, `session_metrics`, into the plugins directory.

```
/var/www/ood/apps/plugins
│── session_metrics
│    ├── initializers
│    ├── lib
│    └── views

```

The attached zip file is an OOD extension to add Job metrics to the session card.
It requires a Slurm job scheduler to work.
Screenshot 2024-11-13 at 14 05 46

[session_metrics.zip](https://github.com/user-attachments/files/18025165/session_metrics.zip)

Fixes #3916

### CODE REVIEW

File: apps/dashboard/config/configuration_singleton.rb [hunk 1]
```diff
@@ -74,8 +74,9 @@ def string_configs
       :rclone_extra_config            => nil,
       :default_profile                => nil,
       :project_size_timeout           => '15',
-      :novnc_default_compression   => '6',
-      :novnc_default_quality       => '2'
+      :novnc_default_compression      => '6',
+      :novnc_default_quality          => '2',
+      :plugins_directory              => '/var/www/ood/apps/plugins'
```
Comment (johrstrom) [inline apps/dashboard/config/configuration_singleton.rb:79]: This should probably be `/etc/ood/config/plugins`. Stuff in `/var/www/ood` will be overwritten on updates.

File: apps/dashboard/config/application.rb [hunk 2]
```diff
@@ -46,5 +46,17 @@ class Application < Rails::Application
...
+
+    # Enable installed plugins only if configured by administrator
+    plugins_dir = Pathname.new(::Configuration.plugins_directory)
+    if plugins_dir.directory?
+      plugins_dir.children.select(&:directory?).each do |installed_plugin|
+        next unless installed_plugin.readable?
+
+        config.paths["config/initializers"] << installed_plugin.join("initializers").to_s
+        config.autoload_paths << installed_plugin.join("lib").to_s
+        config.paths["app/views"].unshift installed_plugin.join("views").to_s
```
Comment (johrstrom) [inline apps/dashboard/config/application.rb:58]: I think we should likely have additional checks here for root owned files in production. I just want to be extra doubly sure that what we load in production is indeed given by an administrator.

```ruby
        init_dir = installed_plugin.join("initializers")
        safe_load = init_dir.children.all? { |f| File.stat(f.to_s).uid.zero? }
        config.paths["config/initializers"] << init_dir if safe_load
```

@treydock any additional thoughts here on loading trusted files?

Comment (abujeda) [reply apps/dashboard/config/application.rb:58]: I understand the need to make sure that the configuration is given by an admin, but why is this different fromthe ones we use for `load_external_config?`, `::Configuration.config_root` (/etc/ood/config/apps/dashboard)?

Comment (johrstrom) [reply apps/dashboard/config/application.rb:58]: We should patch those too, we just haven't.

File: apps/dashboard/config/configuration_singleton.rb (hunk 1 above)
Comment (abujeda) [reply apps/dashboard/config/configuration_singleton.rb]: fixed

File: apps/dashboard/config/application.rb (hunk 2 above)
Comment (abujeda) [reply apps/dashboard/config/application.rb:60]: I made a change to verify that the plugin folders are owned by root when in production

**Pull request merged**